File: exercises/week02/src/solid/FireSensor.java
package solid;

import java.util.Random;

/**
 * An implementation of sensors for fire
 */
class FireSensor implements BatteryPoweredSensor {
    private final String location;
    private int batteryPercentage;

    /**
     * A random number generator to use to simulate a random chance of fire being detected.
     */
    private final Random random;


    public FireSensor(String location) {
        this.location = location;
        random = new Random();
        batteryPercentage = 100;
    }

    @Override
    public boolean isTriggered() {
        batteryPercentage -= 10;
        return random.nextInt(100) < 5;
    }

    @Override
    public String getLocation() {
        return location;
    }

    @Override
    public String getSensorType() {
        return "Fire sensor";
    }

    @Override
    public Type getType() {
        return Type.HAZARD;
    }

    @Override
    public int getBatteryPercentage() {
        return batteryPercentage;
    }
}

File: exercises/week02/src/solid/FireSensor.scala
import scala.util.Random

class FireSensor extends Sensor {

  var battery = 100.0

  override def isTriggered: Boolean = {
    val random = new Random()
    battery -= 10
    random.nextInt(100) < 5
  }

  override def getLocation: String = "unknown location"

  override def getSensorType: String = "fire sensor"

  override def getBatteryPercentage: Double = battery
}
